topics/threaded-io.md

```diff
@@ -0,0 +1,127 @@
+                                                    | attr: *conn     |
+                                                    +-----------------+
+                                                    | attr: querybuf  |
+                                                    +-----------------+
+                                                    | attr: buf       |
+                                                    +-----------------+
+                                                    | ...             |
+                                                    +-----------------+
+
+
+```
+
+### Dispatch read/write tasks to I/O threads
+So when the `server.clients_pending_read` is ready (same for the writing process), we need to dispatch them to the I/O threads which are waiting.
+
+Redis prepared a global variable `io_threads_list` to store the task. `io_threads_list` is a list type variable with thread number length (configured by `io-threads n`). Its item value is a list as well which each sub-item stores the `client` object pointer.
+
+`io_threads_list[i]` represents the task list (client list) the i-th thread should handle. Redis dispatch the pending client using **Round-Robin** algorithm. So when we configured **4 I/O threads** (thread 0 to thread 3), and have **7 clients** (client 0 to client6) in `server.clients_pending_read` list, they will be dispatched like:
+```
+io_threads_list
++---------------+------------+------------+------------+
+|   thread 0    |  thread 1  |  thread 2  |  thread 3  |
+| (main thread) |(I/O thread)|(I/O thread)|(I/O thread)|
++---------------+------------+------------+------------+
+      \/              \/           \/            \/        
+      \/              \/           \/            \/             
+      \/              \/           \/            \/        
+  +----------+   +----------+  +----------+  +----------+
+  | client 0 |   | client 1 |  | client 2 |  | client 3 |
+  +----------+   +----------+  +----------+  +----------+
+  | client 4 |   | client 5 |  | client 6 |
+  +----------+   +----------+  +----------+
+
+```
+
+Each thread can iterate over the clients they should read, do read action, and store the result to `client.querybuf` attribute. The main thread will also handle part of the pending tasks after dispatching. In this example, thread 0, which is the main thread, will read through client 0 and then client 4, thread 1 will handle client 1 and client 5, and so on.
+
+### Blocking and Locking
+There will be a racing condition if the I/O thread keeps monitoring and editing the `io_threads_list[i]` list. So Redis prepared another global variable named `io_threads_pending`. It's a thread number length list and stores an integer value representing the number of the remaining task that thread i should handle. When the main thread finished dispatching task to `io_threads_list`, it set `io_threads_pending[i]` to the actual task count.
```

Comment:
Should we use this section to describe briefly that the main thread signals the io-threads for pending work by setting an atomic variable `io_threads_pending`, then the threads move from busy-waiting to a stage in which they process the pending writes/reads and during that time the main thread is busy waiting for that work to be done and after that the io-threads are put into a busy-wait state again? I believe that the code analysis is too much for the docs? IMHO it's better to describe what the code does... WDYT?

Reply:
Actually the codes might be helpful for me. But yes the document should be more concise and easy to understand for all people who read it.